R/pareto_smooth.R

```diff
@@ -0,0 +1,326 @@
+#'   If `TRUE`, returns x with smoothed tail(s). If `FALSE`, acts the
+#'   same as `pareto_khat`. Default is `TRUE`.
+#' @return List containing vector of smoothed values and vector of
+#'   pareto smoothing diagnostics.
+#'
+#' @examples
+#' mu <- extract_variable_matrix(example_draws(), "mu")
+#' pareto_smooth(mu)
+#'
+#' d <- as_draws_rvars(example_draws("multi_normal"))
+#' pareto_smooth(d$Sigma)
+#' @export
+pareto_smooth <- function(x, ...) UseMethod("pareto_smooth")
+
+#' @rdname pareto_smooth
+#' @export
+pareto_smooth.rvar <- function(x, ...) {
+    unlist(summarise_rvar_by_element_with_chains(x, pareto_smooth.default, ...), recursive = FALSE)
+}
+
+#' @rdname pareto_smooth
+#' @export
+pareto_smooth.default <- function(x,
+                                  tail = c("both", "right", "left"),
+                                  r_eff = NULL,
+                                  ndraws_tail = NULL,
+                                  extra_diags = FALSE,
+                                  verbose = FALSE,
+                                  return_smoothed = TRUE,
+                                  ...) {
+
+  # check for infinite or na values
+  if (should_return_NA(x)) {
+    return(NA_real_)
+  }
+
+  tail <- match.arg(tail)
+  S <- length(x)
+  
+  # automatically calculate relative efficiency
+  if (is.null(r_eff)) {
+    r_eff <- ess_basic(x) / S
+  }
+
+  # automatically calculate tail length
+  if (is.null(ndraws_tail)) {
+    ndraws_tail <- ps_tail_length(S, r_eff)
+  }
+  
+  if (tail == "both") {
+    
+    if (ndraws_tail > S / 2) {
+      warning("Number of tail draws cannot be more than half ",
+              "the length of the draws if both tails are fit, ",
+              "changing to ", S / 2, ".")
+      ndraws_tail <- S / 2
+    }
+
+    if (ndraws_tail < 5) {
+      warning("Number of tail draws cannot be less than 5. ",
+              "Changing to ", 5, ".")
+      ndraws_tail <- 5
+    }
+
+    # left tail
+    smoothed <- .pareto_smooth_tail(
+      x,
+      ndraws_tail = ndraws_tail,
+      tail = "left"
+    )
+    left_k <- smoothed$k
+
+    # right tail
+    smoothed <-.pareto_smooth_tail(
+      x = smoothed$x,
+      ndraws_tail = ndraws_tail,
+      tail = "right"
+    )
+    right_k <- smoothed$k
+
+    k <- max(left_k, right_k)
+    x <- smoothed$x
+  } else {
+    
+    smoothed <- .pareto_smooth_tail(
+      x,
+      ndraws_tail = ndraws_tail,
+      tail = tail
+    )
+    k <- smoothed$k
+    x <- smoothed$x
+  }
+
+  diags <- list(khat = k)
+
+  if (extra_diags) {
+    ext_diags <- .pareto_smooth_extra_diags(k, S)
+    diags <- c(diags, ext_diags)
+  }
+
+  if (verbose) {
+    if (!extra_diags) {
+      diags <- .pareto_smooth_extra_diags(diags$khat, length(x))
+    }
+    pareto_k_diagmsg(
+      diags = diags
+    )
+  }
+  if (return_smoothed) {
+    out <- list(x = x, diagnostics = diags)
+  } else {
+    out <- diags
+  }
+  return(out)
+}
+
+.pareto_smooth_tail <- function(x,
```

Comment:
add at least one or two lines of comments to explain what these internal functions do (same with other internal functions unless they are trivial)

Reply:
I added comments for all internal functions.